x/datadeal/abci.go

```diff
@@ -1,13 +1,72 @@
 package datadeal
 
 import (
+	"strconv"
+
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/medibloc/panacea-core/v2/x/datadeal/keeper"
 	"github.com/medibloc/panacea-core/v2/x/datadeal/types"
 	oracletypes "github.com/medibloc/panacea-core/v2/x/oracle/types"
 )
 
 func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
+	keeper.IterateClosedDataVerificationQueue(ctx, ctx.BlockHeader().Time, func(dataSale *types.DataSale) bool {
+
+		keeper.RemoveDataVerificationQueue(ctx, dataSale.DealId, dataSale.VerifiableCid, dataSale.VotingPeriod.VotingEndTime)
+		iterator := keeper.GetDataVerificationVoteIterator(ctx, dataSale.DealId, dataSale.VerifiableCid)
+
+		defer iterator.Close()
+
+		oracleKeeper := keeper.GetOracleKeeper()
+
+		tallyResult, err := oracleKeeper.Tally(
+			ctx,
+			iterator,
+			func() oracletypes.Vote {
+				return &types.DataVerificationVote{}
+			},
+			func(vote oracletypes.Vote) error {
+				return keeper.RemoveDataVerificationVote(ctx, vote.(*types.DataVerificationVote))
+			},
+		)
+
+		if err != nil {
+			panic(err)
+		}
+
+		if tallyResult.IsPassed() {
+			dataSale.Status = types.DATA_SALE_STATUS_DELIVERY_VOTING_PERIOD
+			if dataSale.VerifiableCid != string(tallyResult.ConsensusValue) {
+				panic("invalid verifiable CID consensus value")
+			}
+
+			keeper.AddDataDeliveryQueue(ctx, dataSale.VerifiableCid, dataSale.DealId, oracleKeeper.GetVotingPeriod(ctx).VotingEndTime)
+		} else {
+			dataSale.Status = types.DATA_SALE_STATUS_VERIFICATION_FAILED
+		}
+
+		dataSale.VerificationTallyResult = tallyResult
+
+		if err := keeper.SetDataSale(ctx, dataSale); err != nil {
+			panic(err)
+		}
+
+		ctx.EventManager().EmitEvents(sdk.Events{
+			sdk.NewEvent(
+				types.EventTypeDataVerificationVote,
+				sdk.NewAttribute(types.AttributeKeyVoteStatus, types.AttributeValueVoteStatusEnded),
+				sdk.NewAttribute(types.AttributeKeyVerifiableCID, dataSale.VerifiableCid),
+				sdk.NewAttribute(types.AttributeKeyDealID, strconv.FormatUint(dataSale.DealId, 10)),
+			),
+			sdk.NewEvent(
+				types.EventTypeDataDeliveryVote,
+				sdk.NewAttribute(types.AttributeKeyVoteStatus, types.AttributeValueVoteStatusStarted),
```

Comment:
This event should be emitted only when the verification vote is passed.

Reply:
So, this event should be in the `if tallyResult.IsPassed()` condition?
IMHO, I think that this event emission also means the verification vote is passed.

Reply:
Right. let's move this emission inside the if statement.

Reply:
Fixed in [5790d95](https://github.com/medibloc/panacea-core/pull/449/commits/5790d955cd663d845094563b94667d232c5dadaa).